### PR TITLE
Gradle plugin improvements

### PR DESCRIPTION
**Pull request by roguexz**
* Split the Gradle integration in to two pieces - tooling-api & plugin implementation. This paves the way for improving the Arquillian integration.
* Update the plugin name from "wildfly-swarm" to "thorntail"
* Refactored the dependency resolution code. This should help with fixing / reconciling dependency resolution issues between the way Maven & Gradle work.
* Improved the Thorntail configuration to accept options similar to the Maven plugin.
* Verified the plugin's capability on a multi-module Gradle project.

- [x] Have you followed the guidelines in our [Contributing](http://wildfly-swarm.io/community/contributing/) document?
- [ ] [v2] Have you created a [JIRA](https://issues.jboss.org/browse/THORN) and used it in the commit message?
- [ ] [v4] Have you created a [GitHub Issue](https://github.com/thorntail/thorntail/issues) and used it in the commit message?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/thorntail/thorntail/pulls) for the same issue?
- [x] Have you built the project locally prior to submission with `mvn clean install`?

-----

### CODE REVIEW

**Comment by roguexz** [comment]
@kenfinnigan @Ladicek 
This is the first (complete) PR that restructures the Gradle plugin. The examples that I have verified are [here](https://github.com/roguexz/thorntail-examples/tree/master/gradle-examples).

In subsequent PRs I will be improving the Arquillian integration support.

Please provide your feedback.

File: build-parent/pom.xml [hunk 1]
```diff
@@ -77,7 +77,7 @@
 
     <version.flyway>4.2.0</version.flyway>
 
-    <version.gradle>2.4</version.gradle>
+    <version.gradle>3.0</version.gradle>
```
Comment (roguexz) [inline build-parent/pom.xml:80]: The Gradle tooling dependency was already at 3.0. I am ensuring that we make use of the same version across.

File: plugins/gradle/gradle-plugin/pom.xml [hunk 2]
```diff
@@ -31,6 +31,12 @@
         </configuration>
       </plugin>
     </plugins>
+    <resources>
+      <resource>
+        <directory>${project.basedir}/src/main/resources</directory>
+        <filtering>true</filtering>
+      </resource>
+    </resources>
```
Comment (roguexz) [inline plugins/gradle/gradle-plugin/pom.xml:39]: Filtering the resources for embedding the Thorntail project version in to the properties file.

**Comment by sberyozkin** [comment]
Hi, would it make sense, given you are doing the major restructuring, update to the latest Gradle 4.10 ?

**Comment by sberyozkin** [comment]
Can you please check that https://github.com/thorntail/thorntail-examples/tree/master/gradle works

**Comment by roguexz** [comment]
@sberyozkin 

*Regarding updating to Gradle 4.10*
This is going to be a challenge since the core-api libraries (of Gradle) are not published to any of the repositories. What this means is that if we wish to build the plugin as part of the Maven build system (with minimal changes), we will need to use a minimum Gradle version of 3.0

Having said that, I have verified that the [examples](https://github.com/roguexz/thorntail-examples/tree/master/gradle-examples) works with [4.10](https://github.com/roguexz/thorntail-examples/blob/master/gradle-examples/gradle/wrapper/gradle-wrapper.properties#L6).

I have included a multi-module example as well that show cases the use of the plugin.

The next PR that I will be raising (already tested and ready to ship), updates the Gradle Arquillian adapter to use the existing tooling model for fetching all the project dependencies. What that means, is that we can make use of Gradle's [IMPROVED_POM_SUPPORT](https://docs.gradle.org/current/userguide/managing_transitive_dependencies.html) to build Thorntail based projects.

**Comment by roguexz** [comment]
Below is the simplified view of the changes that are being pushed as part of this PR.
* ThorntailConfiguration: is the interface that defines the plugin's configuration options. This interface is used for serializing data across the wire to the Arquillian adapter (next PR).
* DependencyDescriptor: is the interface that defines GAV coordinates and allows me to serialize data across the wire.
* ThorntailApplicationModel: interface that encapsulates the ThorntailConfiguration reference and provides additional helper functions. This acts as the base for sending additional details across the wire.
* GradleDependencyResolutionHelper: Provides capabilities for resolving dependencies and fraction details, etc.
* GradleArtifactResolvingHelper, PackageTask, PackagePlugin: Moved from the previous location to the new location.

Let me know if you would like to know more about the flow of interactions between the various classes.

class-design

**Comment by roguexz** [comment]
There is potential to collapse some of these classes. I shall do that as part of the next PR (when I update the Arquillian adapter).

**Comment by sberyozkin** [comment]
Thanks, from a high, non Gradle expert level it all looks like a very good effort, refreshing and improving the existing Gradle plugin is good. 

I'd appreciate if you could confirm that the existing simple Gradle demo which I linked still works. Or may be it would be better for you to replace that demo with a new set of your examples in the next PR ?

**Comment by roguexz** [comment]
@sberyozkin

I'm sorry for not clarifying it earlier. Yes, it works. I renamed the project to 'jaxrs-cdi' in the associated fork of the examples project.

**Comment by roguexz** [comment]
@sberyozkin Ok .. Let me make one correction to that above statement. There is a breaking change in this PR. I have renamed the plugin and the configuration name from `swarm` to `thorntail`. So,

previously the build script would look like,

```
apply plugin: 'wildfly-swarm'
...
swarm {
  properties {
    swarm.http.port = 8181
  }
}
```

after my change it would look like,

```
apply plugin: 'thorntail'
...
thorntail {
  properties {
    swarm.http.port = 8181
  }
}
```

So, looks like the examples project (as of now in master) does not compile (because the plugin was renamed to [`thorntail`](https://github.com/thorntail/thorntail-examples/blob/master/gradle/build.gradle#L16) though there is no such plugin being exposed by the code). But I will be raising a PR to update the examples project as well.

**Comment by roguexz** [comment]
Raised a [PR for the examples project](https://github.com/thorntail/thorntail-examples/pull/191) as well.

**Comment by roguexz** [comment]
@Ladicek Any idea why the Linux build hasn't kicked off yet?

**Comment by Ladicek** [comment]
Not sure why, but we seem to have an issue in communication between Jenkins and GitHub, as I've now seen multiple PRs not having a PR test reported in GitHub, yet having a green Jenkins build. This PR was also tested successfully: https://ci.wildfly-swarm.io/view/Pull%20Requests/job/thorntail-pull-request-linux/711/

**Comment by sberyozkin** [comment]
@roguexz Thanks, this is all looking good now.

**Review by sberyozkin — APPROVED** [review]
@Ladicek, @kenfinnigan I think it should be safe enough to merge it before 2.2.1 is out, will merge later today if there will be no objections

**Pull request merged**